core/test/Lender.t.sol

```diff
@@ -3,21 +3,170 @@ pragma solidity ^0.8.15;
+        // Give this test contract some shares
+        deal(address(asset), address(lender), 2e18);
+        lender.deposit(2e18, address(this));
+
+        // Borrow some tokens (so that interest will actually accrue)
+        hoax(lender.FACTORY());
+        lender.whitelist(address(this));
+        lender.borrow(1e18, address(this));
+
+        // Mock interest model
+        accrualFactor = accrualFactor % 0.1e12;
+        vm.mockCall(
+            address(lender.interestModel()),
+            abi.encodeWithSelector(InterestModel.getAccrualFactor.selector, 13, 0.5e18),
+            abi.encode(accrualFactor)
+        );
+
+        uint256 newInventory = 1e18 + FixedPointMathLib.mulDivDown(1e18, accrualFactor + 1e12, 1e12);
+        uint256 interest = newInventory - 2e18;
+        uint256 reserves = interest / lender.reserveFactor();
+
+        uint256 epsilon = 2;
+
+        // Initially, the `lender` holds 1e18 and has lent out 1e18
+        assertEqDecimal(lender.totalAssets(), 2e18, 18);
+        // Now skip forward in time, but don't modify `lender` state yet
+        skip(13);
+        // `totalAssets` uses view-only methods to compute interest, so it should have updated
+        assertEqDecimal(lender.totalAssets(), newInventory, 18);
+        // But `balanceOfUnderlyingStored` just reads from storage, so it should still have old value
+        assertEqDecimal(lender.balanceOfUnderlyingStored(address(this)), 2e18, 18);
+        // Now actually store interest updates
         lender.accrueInterest();
+        // Both `totalAssets` and `balanceOfUnderlyingStored` should have updated now
+        assertEqDecimal(lender.totalAssets(), newInventory, 18);
+        assertLeDecimal(
+            stdMath.delta(
+                lender.balanceOfUnderlyingStored(address(this)),
+                newInventory - reserves
+            ),
+            epsilon,
+            18
+        );
+        // Make sure `borrowIndex` is just as precise as `accrualFactor`
+        if (accrualFactor > 0) assertGt(lender.borrowIndex(), 1e12);
+
+        assertEq(lender.lastAccrualTime(), block.timestamp);
+        assertLeDecimal(
+            stdMath.delta(
+                lender.balanceOfUnderlying(lender.RESERVE()),
+                reserves
+            ),
+            epsilon,
+            18
+        );
+
+        vm.clearMockedCalls();
+    }
+
+    function test_cannotInflateSharePrice(uint128 amount) public {
+        deal(address(asset), address(lender), 1e18);
+        lender.deposit(1e18, address(12345));
+
+        uint256 balance = lender.balanceOfUnderlying(address(12345));
+        deal(address(asset), address(lender), amount);
+
+        assertEq(lender.balanceOfUnderlying(address(12345)), balance);
+    }
+
+    function test_cannotDepositWithoutERC20Transfer(uint112 amount, address to) public {
+        if (amount == 0) amount++;
+
+        deal(address(asset), address(lender), amount - 1);
+        vm.expectRevert(bytes(""));
```

Comment:
We should specify the revert

Reply:
This is as specific as we can get right now, since no real error message is emitted. `bytes("")` at least tells it to expect an empty reason string, as opposed to _any_ reason string.